src/components/ReplaceServerDetails.js

```diff
@@ -0,0 +1,321 @@
+          osInstallUsername: responseData['username']
+        });
+      });
+  }
+
+  isFormTextInputValid() {
+    let isAllValid = true;
+    let values = Object.values(this.allInputsStatus);
+    isAllValid =
+      (values.every((val) => val === INPUT_STATUS.VALID));
+    //also check osInstall checked and need password
+    if(this.state.isInstallOsChecked) {
+      isAllValid = isAllValid && this.state.osInstallPassword && this.state.osInstallPassword !== '';
+    }
+
+    return isAllValid;
+  }
+
+  updateFormValidity = (props, isValid) => {
+    this.allInputsStatus[props.inputName] = isValid ? INPUT_STATUS.VALID : INPUT_STATUS.INVALID;
+    this.setState({isFormValid: this.isFormTextInputValid()});
+  }
+
+  handleDone = () => {
+    if(!this.data.uid) {
+      this.data.uid = genUID();
+    }
+    this.data['mac-addr'] = this.state.replaceData.get('mac-addr');
+    this.data['ilo-ip'] = this.state.replaceData.get('ilo-ip');
+    this.data['ilo-user'] = this.state.replaceData.get('ilo-user');
+    this.data['ilo-password'] = this.state.replaceData.get('ilo-password');
+    //if it is an existing server discovered use that uid
+    this.data['uid'] = this.state.replaceData.get('uid');
+    let theProps = {
+      wipeDisk : this.state.isWipeDiskChecked,
+      installOS : this.state.isInstallOsChecked
+    };
+    if(this.state.isInstallOsChecked) {
+      theProps.osInstallUsername = this.state.osInstallUsername;
+      theProps.osInstallPassword = this.state.osInstallPassword; //for now TODO
+    }
+    //picked from available server
+    if(this.state.selectedServerId) {
+      theProps.selectedServerId = this.state.selectedServerId;
+    }
+
+    this.props.doneAction(this.data, theProps);
+  }
+
+  handleCancel = () => {
+    this.props.cancelAction();
+  }
+
+  handleInputChange = (e, isValid, props) => {
+    //deselect the selected available server when
+    //user start entering values by himself
+    this.state.selectedServerId = '';
+    this.updateFormValidity(props, isValid);
+    let value = e.target.value;
+    if (isValid) {
+      if(props.inputName !== 'osInstallPassword') {
+        let newData = this.state.replaceData.set(props.inputName, value);
+        this.setState({replaceData: newData});
+      }
+      else {
+        this.setState({osInstallPassword: value}, () => {
+          this.setState({isFormValid: this.isFormTextInputValid()});
+        });
+      }
+    }
+  }
+
+  updateAutoFillFormValidity = (newJSData) => {
+    for (let key in newJSData) {
+      if(newJSData[key]) {
+        this.allInputsStatus[key] = INPUT_STATUS.VALID;
+      }
+      else {
+        this.allInputsStatus[key] = INPUT_STATUS.INVALID;
+      }
+    }
+    this.setState({isFormValid: this.isFormTextInputValid()});
+  }
+
+  handleOsPasswordChange = (e) => {
+    const password = e.target.value;
+    this.setState({osInstallPassword: password}, () => {
+      this.setState({isFormValid: this.isFormTextInputValid()});
```

Comment:
Remove. 
Not only is this unnecessary (as mentioned above), this code is actually wrong. this.setState executes in a promise, and when that promise returns, then rendering can begin. But the callback here is another call to setState which immediately returns a promise, so `isFormValid` is likely to not be set until *after* rendering, which is wrong.

Reply:
will update to simplify and clean up